### PR TITLE
Reshape the LocaleFallbacker API to have the WithConfig intermediate

### PR DESCRIPTION
**Pull request by sffc**
Toward #2683

### CODE REVIEW

File: components/locid_transform/src/fallback/mod.rs [hunk 1]
```diff
@@ -350,32 +355,20 @@ impl LocaleFallbacker {
...
+    /// Associates a configuration with this fallbacker.
+    #[inline]
     pub fn for_config(&self, config: LocaleFallbackConfig) -> LocaleFallbackerWithConfig {
         self.as_borrowed().for_config(config)
     }
 
-    #[doc(hidden)]
+    /// Derives a configuration from a [`DataKey`] and associates it
+    /// with this fallbacker.
+    #[inline]
```
Comment (robertbastian) [inline components/locid_transform/src/fallback/mod.rs:366]: I had this as `doc(hidden)` because a `DataKey` is a data pipeline implementation detail, and not something that should necessarily be on `locid_transform`'s API. Clients can do `key.into()` themselves.

Comment (robertbastian) [reply components/locid_transform/src/fallback/mod.rs:366]: > in most use cases, you start with a DataKey

Disagree. In most use cases *we* start with a `DataKey`, but users will use fallback for other purposes than building a fallback provider.

Comment (sffc) [reply components/locid_transform/src/fallback/mod.rs:366]: Also `.into()` is not `const`

Comment (sffc) [reply components/locid_transform/src/fallback/mod.rs:366]: ok done

**Review by robertbastian — DISMISSED** [review]

File: components/locid_transform/src/fallback/mod.rs [hunk 2]
```diff
@@ -386,22 +380,9 @@ impl LocaleFallbacker {
 }
 
 impl<'a> LocaleFallbackerBorrowed<'a> {
-    /// Creates an iterator based on a [`Locale`].
-    ///
-    /// When first initialized, the locale is normalized according to the fallback algorithm.
-    ///
-    /// For performance reasons, internally an [`icu_provider::DataLocale`] is used.
-    ///
-    /// [`Locale`]: icu_locid::Locale
```
Comment (robertbastian) [inline components/locid_transform/src/fallback/mod.rs:395]: Why did you change this?

Comment (sffc) [reply components/locid_transform/src/fallback/mod.rs:395]: You just noticed? :) because we use DataLocale basically everywhere, in all our constructors, and this is a place where we actually need it. As discussed previously, I dislike `Into` bounds because it creates generics at call sites when people should really just call `.into()` for the same behavior.

Comment (robertbastian) [reply components/locid_transform/src/fallback/mod.rs:395]: I'm not talking about the general concept of using `DataLocale` here, I'm asking why you removed the docs that explain that it's fine if you have a `Locale`, but we use `DataLocale` for performance reasons.

Comment (sffc) [reply components/locid_transform/src/fallback/mod.rs:395]: Discussed and revised docs.

**Review by robertbastian — APPROVED** [review]
